draft-ietf-taps-interface.md

```diff
@@ -1780,38 +1772,170 @@ that the other endpoint is done sending on a connection.
+framer := NewMessageFramer()
+framer.setEventHandlers()
+~~~
+
+## Adding Message Framers to Connections
 
-{messageData} := Deframer.Deframe(OctetStream)
+The Message Framer object can then be added to one or more Preconnections
+to run on top of transport protocols. Multiple Framers may be added.
+
+~~~
+Preconnection.AddFramer(framer)
+~~~
+
+When sending Messages, applications can add specific Message
+values to a MessageContext ({{message-props}}) that is intended for a Framer.
+This can be used, for example, to set the type of a Message for a TLV protocol.
+The namespace of values is custom for each unique Message Framer.
+
+~~~
+messageContext := NewMessageContext()
+messageContext.add(framer, key, value)
+Connection.Send(messageData, messageContext)
+~~~
+
+When an application receives a MessageContext in a Receive event,
+it can also look to see if a value was set by a specific Message Framer.
+
+~~~
+messageContext.get(framer, key) -> value
+~~~
+
+## Message Framer Lifetime
+
+When a Connection establishment attempt begins, an event is delivered to
+notify the framer that a new Connection is being created and has a new
+instance of a Message Framer. Similarly, a stop event is delivered when a
+Connection is being torn down.
+
+~~~
+FramerInstance -> Start()
+FramerInstance -> Stop()
 ~~~
 
+When Message Framer gets a start message, it sets up its state and then indicates
+to the connection that it is ready to handle sending Messages by calling `ready` on
+the FramerInstance. This call can be delayed in case the protocol needs to do some
+simple handshake exchange before handling application Messages.
+
+~~~
+instance.ready()
+~~~
+
+At any time if the Message Framer encounters a fatal error, it can also cause the Connection
+to fail and provide an error.
+
+~~~
+instance.failed(Error)
+~~~
+
+## Sender-side Message Framing {#send-framing}
+
+Message Framers define an event that gets called on an instance of a
+Framer to handle a new Message being sent by the application.
+
+~~~
+FramerInstance -> NewSentMessage<messageData, messageContext, endOfMessage>
+~~~
+
+Upon receiving this event, a Message Framer implementation is responsible for
+performing any necessary transformations and sending the resulting data to the next
+protocol. Implementations SHOULD ensure that there is a way to pass the original data
+through without copying to improve performance.
+
+~~~
+instance.send(Data)
+~~~
+
+To provide an example, a simple protocol that adds a length as a header would receive
+the `NewOutputMessage` event, create a data representation of the length of the Message
+data, and then send a data that is the concatenation of the length header and the original
+Message data.
+
+## Receiver-side Message Framing {#receive-framing}
+
+In order to parse an received flow of data into Messages, the Message Framer
+is first notified whenever new data is available to parse.
+
+~~~
+FramerInstance -> HandleReceivedData<>
```

Comment:
This example broken:
- FramerInstance is the receiver of the event, not the sender
- We have no other events with "handle" in the name
- Arguments (Data + Context are missing)

Reply:
No, FramerInstance is not the receiver of the event, it *is* the sender. It's the handle to the representation of the framer within the Connection that you interact with. It's your state. It says: your framer now has data available.

The data for new inbound data should not be provided in this event—the implementation of the framer needs to parse the data out. The data may not be in a form that is efficiently consumable as a single data object (it's an entire stream).

Reply:
Okay - I got it. Also makes sense not to specify who receives this Event, as it could be another Framer or the connection object to forward it to the application.

Can we try to make this event match the Received/ReceivedPartial event somehow? 
This would look much more consistent and allows to directly send it from the Framer.